libs/Microsoft.MixedReality.WebRTC.Native/include/interop_api.h

```diff
@@ -329,93 +450,135 @@ struct PeerConnectionConfiguration {
+/// time.
 MRS_API void MRS_CALL mrsPeerConnectionRegisterLocalSdpReadytoSendCallback(
-    PeerConnectionHandle peerHandle,
+    mrsPeerConnectionHandle peerHandle,
     PeerConnectionLocalSdpReadytoSendCallback callback,
     void* user_data) noexcept;
 
-/// Register a callback fired when an ICE candidate message is ready to be sent
-/// via the signaling service to a remote peer.
+/// Register a callback invoked when an ICE candidate message is ready to be
+/// sent via the signaling service to a remote peer. Only one callback can be
+/// registered at a time.
 MRS_API void MRS_CALL mrsPeerConnectionRegisterIceCandidateReadytoSendCallback(
-    PeerConnectionHandle peerHandle,
+    mrsPeerConnectionHandle peerHandle,
     PeerConnectionIceCandidateReadytoSendCallback callback,
     void* user_data) noexcept;
 
-/// Register a callback fired when the ICE connection state changes.
+/// Register a callback invoked when the ICE connection state changes. Only one
+/// callback can be registered at a time.
 MRS_API void MRS_CALL mrsPeerConnectionRegisterIceStateChangedCallback(
-    PeerConnectionHandle peerHandle,
+    mrsPeerConnectionHandle peerHandle,
     PeerConnectionIceStateChangedCallback callback,
     void* user_data) noexcept;
 
 /// Register a callback fired when a renegotiation of the current session needs
 /// to occur to account for new parameters (e.g. added or removed tracks).
 MRS_API void MRS_CALL mrsPeerConnectionRegisterRenegotiationNeededCallback(
-    PeerConnectionHandle peerHandle,
+    mrsPeerConnectionHandle peerHandle,
     PeerConnectionRenegotiationNeededCallback callback,
     void* user_data) noexcept;
 
-/// Register a callback fired when a remote media track is added to the current
+/// Register a callback fired when a remote audio track is added to the current
 /// peer connection.
-MRS_API void MRS_CALL mrsPeerConnectionRegisterTrackAddedCallback(
-    PeerConnectionHandle peerHandle,
-    PeerConnectionTrackAddedCallback callback,
+/// Note that the arguments include some object handles, which each hold a
+/// reference to the corresponding object and therefore must be released, even
+/// if the user does not make use of them in the callback.
+MRS_API void MRS_CALL mrsPeerConnectionRegisterAudioTrackAddedCallback(
+    mrsPeerConnectionHandle peerHandle,
+    PeerConnectionAudioTrackAddedCallback callback,
     void* user_data) noexcept;
 
-/// Register a callback fired when a remote media track is removed from the
+/// Register a callback fired when a remote audio track is removed from the
 /// current peer connection.
-MRS_API void MRS_CALL mrsPeerConnectionRegisterTrackRemovedCallback(
-    PeerConnectionHandle peerHandle,
-    PeerConnectionTrackRemovedCallback callback,
+/// Note that the arguments include some object handles, which each hold a
+/// reference to the corresponding object and therefore must be released, even
+/// if the user does not make use of them in the callback.
```

Comment:
Why does the callback implementation have to be made responsible of releasing the reference?

nit: I'd rather mention the arguments that need to be released specifically.

Reply:
I see only two lifecycle options:
- An object is explicitly created/destroyed via its handle with a specific pair of API calls, and stays alive between those two points, whether or not it is in use. This is usually how handles are used in a standard C API.
- An object is reference-counted, and that reference count is used to drive the destruction, in which case each user must keep a reference to it. This is generally what a C++ API uses, along with a helper like a shared pointer.

Because we started originally by the latter, now each handle holds a reference, and passing through the interop layer a handle without reference makes it very confusing, as now you have handles that do have a reference (_e.g._ returned from a "create" method) and handles that do not (_e.g._ passed as argument to callback), which makes things even more confusing and error-prone in my opinion for the developers.

I do appreciate the current approach is also a bit unusual and so possibly confusing. But what is a better alternative? Using non-reference-counted plain C handles?

Reply:
> But what is a better alternative? Using non-reference-counted plain C handles?

Discussed IRL. Sounds like this is a better alternative.